lib/math/enum.ex

```diff
@@ -82,24 +83,58 @@ defmodule Math.Enum do
 
   def median(collection) do
     count = Enum.count(collection)
+    mid_point = div(count, 2)
 
     cond do
       count == 0 ->
         nil
 
       # Middle element exists
-      rem(count, 2) == 1 ->
-        Enum.sort(collection) |> Enum.at(div(count, 2))
+      Integer.is_odd(count) ->
+        collection
+        |> Enum.sort()
+        |> Enum.fetch!(mid_point)
 
       true ->
-        # Take two middle-most elements.
-        sorted_collection = Enum.sort(collection)
-
-        [
-          Enum.at(sorted_collection, div(count, 2)),
-          Enum.at(sorted_collection, div(count, 2) - 1)
-        ]
+        collection
+        |> Enum.sort()
+        |> Enum.slice((mid_point - 1)..mid_point)
         |> Math.Enum.mean()
     end
   end
+
+  @doc """
+  Calculates the mode of a given collection of numbers.
+
+  Always returns a list. An empty input results in an empty list. Supports bimodal/multimodal collections by returning a list with multiple values.
+
+  ## Examples
+
+      iex> Math.Enum.mode [1, 2, 3, 4, 1]
+      [1]
+      iex> Math.Enum.mode [1, 2, 3, 2, 3]
+      [2, 3]
+      iex> Math.Enum.mode []
+      []
+      # iex> Math.Enum.mode [1, 1.0, 2, 3]
+      # [1]
+      # iex> Math.Enum.mode [1, 1.0, 2, 3] &===/1
+      # [1, 1.0]
+  """
+  @spec mode(Enum.t()) :: Enum.t()
+  def mode(collection)
+
+  def mode(collection) do
+    case Enum.count(collection) do
```

Comment:
Why do you do this extra check? Depending on the enumerable this might require yet another linear iteration across the whole structure, and the functions in the non-zero case handle empty enumerables just fine.

Reply:
`max_by/3` will fail on an empty collection however I might be able to get around this by using `with` or some other method of sorting. 

I'll look into this...

Reply:
`Enum.max_by` has a three-argument version where you can specify what happens in the case of an empty collection.

Reply:
Yeah, never used that before. Just updated it!